---
title: "Exercise 2 - Bioconductor"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/scRNA-seq"
output: 
  html_document:
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  toMessage: false    
---

```{r, setup,include=FALSE}
toMessage <- ifelse(!is.null(params$toMessage),params$toMessage,FALSE)


```

<p>&nbsp;</p>
These exercises are about the simple scRNAseq workflow in [session 2](https://rockefelleruniversity.github.io/https://rockefelleruniversity.github.io/SingleCell_Bootcamp/). 

## Description

We will use data from this [study](https://www.science.org/doi/10.1126/science.aat1699) on renal cancer from human cancer. 

The full dataset is on [ArrayExpress](https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-7407). We will use just one sample for now: 4834STDY7002881.

You can also download it from [DropBox](https://www.dropbox.com/scl/fo/1k4y3fvfbissu1vds5kru/AN-CDAVfqa7H5pJT4wCchj0?rlkey=xi9fvilmqjt5j5er73cbocnui&st=zwlvcf6e&dl=0). 

**Exercise 1** - **Cell Ranger**

- Review the Web Summary Report
- What version of Cell Ranger was used to process this data?
```{r, eval=FALSE,echo=toMessage}
# cellranger-4.0.0
```
- How many cells are there?

- What looks problematic about this dataset?

- What steps could we do to to help things?


**Exercise 2** - **Create a Seurat Object**

- Read the matrix file in and create a Seurat object. Ensure it is properly annotated. 
```{r, eval=FALSE,echo=toMessage}
library(Seurat)

mtx <- Read10X("~/Downloads/4834STDY7002881/GRCh38/")

seu <- CreateSeuratObject(mtx, project="kidney_F6", min.cells=10, min.features=200)
seu[["dset"]] <- "kidney_F6"
seu <- Seurat::RenameCells(seu, add.cell.id="kidney_F6")

```

- Run normalization (log) and scaling. What are the top variable features?

```{r, eval=FALSE,echo=toMessage}

seu <- NormalizeData(seu, normalization.method="LogNormalize")
seu <- FindVariableFeatures(seu, select.method="vst", nfeatures=3000)
seu <- ScaleData(seu)

top10 <- head(VariableFeatures(seu), 10)

temp_plot <- VariableFeaturePlot(seu)
plot2 <- LabelPoints(plot = temp_plot, points = top10, repel = TRUE)
plot2
```

- Do dimension reduction. Don't forget to check the elbow for PCA to get a good number of PCs. Than run some basic clusetering. No need to optimize that. 

```{r, eval=FALSE,echo=toMessage}

library(dplyr)
seu <- RunPCA(seu, assay = "RNA", npcs = 50)
pct <- seu[["pca"]]@stdev / sum(seu[["pca"]]@stdev) * 100
cumu <- cumsum(pct)
co1 <- which(cumu > 90 & pct < 5)[1]
co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = T)[1] + 1
pc <- min(co1, co2)

plot_df <- data.frame(pct = pct, cumu = cumu,rank = 1:length(pct))

ggplot(plot_df, aes(cumu, pct, label = rank, color = rank > pc)) + geom_text() +
  geom_vline(xintercept = 90, color = "grey") +
  geom_hline(yintercept = min(pct[pct > 5]), color = "grey") +
  theme_bw()

seu <- FindNeighbors(seu,dims = 1:pc, reduction = "pca")
seu <- RunUMAP(seu,dims = 1:pc, reduction = "pca")

seu <- FindClusters(seu, resolution = 0.5)
seu[["cluster_byDefault"]] <- seu$seurat_clusters

DimPlot(seu)
```


**Exercise 3** - **Basic QC**

- Review the number of counts and features detected. Is there any asymmetry?

```{r, eval=FALSE,echo=toMessage}

VlnPlot(seu,
        features = c("nCount_RNA","nFeature_RNA"))

FeaturePlot(seu,
        features = c("nCount_RNA","nFeature_RNA"))

```

- Lets check the mitochondrial content. Review the results. 

```{r load_data,include=TRUE, eval=FALSE,echo=toMessage}
library(Seurat)

seu$percent.mt <- PercentageFeatureSet(seu, pattern = "^MT-")

VlnPlot(seu,"percent.mt", group.by ="dset")

FeaturePlot(seu,"percent.mt")  + scale_colour_gradient(low="gray",high="blue")

FeatureScatter(seu, feature1 = "nCount_RNA", feature2 = "percent.mt")

table(seu$percent.mt>10)
```

- Lets run cell cycle checks?

```{r ccPhase_Seurat,include=TRUE,eval=TRUE, eval=FALSE,echo=toMessage}
feat_s <- cc.genes$s.genes
feat_g2m <- cc.genes$g2m.genes

seu <- CellCycleScoring(seu, s.features = feat_s, g2m.features = feat_g2m)

DimPlot(seu, group.by = "Phase")

dat_s <- data.frame(cell_id=Cells(seu), cat="S_Score", Phase=seu$Phase, score=seu$S.Score)

dat_g2m <- data.frame(cell_id=Cells(seu), cat="G2M_Score", Phase=seu$Phase, score=seu$G2M.Score)

dat <- rbind(dat_s, dat_g2m)
dat$Phase <- factor(dat$Phase, levels = c("G1","S","G2M"))

library(scran)

sce <- as.SingleCellExperiment(seu, assay = "RNA")
rowData(sce)$SYMBOL <- rownames(sce)

load("data/ccGene_mouse_human_geneSymbol_ensemblID_20220817.RData")
ccGene_hs <- ccGene_mm_hs$human_symbol
assignments <- cyclone(sce, ccGene_hs, gene.names=rowData(sce)$SYMBOL)

seu[["cyclone_Phase"]] <- assignments$phases
seu[["cyclone_G1Score"]] <- assignments$scores$G1
seu[["cyclone_SScore"]] <- assignments$scores$S
seu[["cyclone_G2MScore"]] <- assignments$scores$G2M

DimPlot(seu, group.by = "cyclone_Phase")

```

- Lets run scrublet. Review the doublet annotation. Are doublets an issue for this dataset?

```{R, eval=FALSE,echo=toMessage}
scr <- reticulate::import("scrublet")
mat <- GetAssayData(seu, assay = "RNA", slot = "counts")
mat <- as.matrix(mat)
scrub <- scr$Scrublet(t(mat))
doublet <- scrub$scrub_doublets()
names(doublet) <- c("doublet_score","doublet")


seu[["doublet_score"]] <- doublet$doublet_score
seu[["doublet"]] <- doublet$doublet


VlnPlot(seu, group.by = "doublet",
        features = c("doublet_score", "nCount_RNA","nFeature_RNA"),
        pt.size = 0)
```

**Exercise 4** - **Subset and Regress**

- Subset the data and run regression. Review the data and think about what features it is appropriate to regress over.

```{r, eval=FALSE,echo=toMessage}
seu_filt <- subset(seu, subset=doublet=="FALSE" & 
                     percent.mt < 10)

# These cells are from a cancer study so there's a good chance there's a high proportion of cycling cells -> Seurat cell cycle assignment seems good. 
pot_conf <- c("percent.mt","S.Score","G2M.Score")
seu_filt <- ScaleData(seu_filt, vars.to.regress = pot_conf)

```

- Rerun PCA, UMAP and clustering. Review the results including QC features. 

```{r, eval=FALSE,echo=toMessage}
  seu_filt <- RunPCA(seu_filt, npcs=30, verbose=FALSE)
  seu_filt <- RunUMAP(seu_filt, reduction = "pca", dims = 1:10, verbose=FALSE)
  seu_filt <- FindNeighbors(seu_filt, reduction = "pca", dims = 1:10, verbose=FALSE)
  seu_filt <- FindClusters(seu_filt, resolution = 0.5, verbose=FALSE)
  
  DimPlot(seu_filt)
  DimPlot(seu_filt, group.by ="Phase")
  FeaturePlot(seu_filt,"percent.mt")  + scale_colour_gradient(low="gray",high="blue")
  
  
```

**Exercise 4** - **Optimize Clustering and Find Markers**

- Optimize Clustering

```{r, eval=FALSE,echo=toMessage}
library(clustree)

seu_filt$default_clusters <- seu_filt$seurat_clusters

reso <- c(0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1)
reso_res <- lapply(1:length(reso), function(x,seu_filt,reso){
  seu_filt <- FindClusters(seu_filt,resolution = reso[x])
  clust <- setNames(seu_filt$seurat_clusters,Cells(seu_filt))
  return(clust)}, seu_filt, reso)
names(reso_res) <- paste0("k",1:length(reso))


k_tab <- do.call(cbind,reso_res)
k_dat <- as.data.frame(k_tab)

clustree(k_dat, prefix = "k", node_colour = "sc3_stability")

seu_filt <- FindClusters(seu_filt, resolution = 0.2)

DimPlot(seu_filt, group.by = "seurat_clusters",label = TRUE,pt.size = 0.2) + NoLegend() + ggtitle("Optimized Clusters")

DimPlot(seu_filt, group.by = "default_clusters",label = TRUE,pt.size = 0.2) + NoLegend() + ggtitle("Default Clusters")


```

- Find Markers
```{r, eval=FALSE,echo=toMessage}
library(magrittr)
markers <- FindAllMarkers(seu_filt, only.pos = TRUE,
                          min.pct = 0.25, logfc.threshold = 0.25)


top_genes <- markers %>% group_by(cluster) %>%
  slice_max(n = 2, order_by = avg_log2FC)
head(top_genes)

DoHeatmap(seu_filt, features = top_genes$gene) + NoLegend()
```


