<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Single-cell RNA sequencing ~ Session 1</title>
    <meta charset="utf-8" />
    <meta name="author" content="https://rockefelleruniversity.github.io/SingleCell_Bootcamp/" />
    <meta name="author" content="brc@rockefeller.edu" />
    <script src="libs/header-attrs-2.29/header-attrs.js"></script>
    <link rel="stylesheet" href="default.css" type="text/css" />
    <link rel="stylesheet" href="metropolisCustom.css" type="text/css" />
    <link rel="stylesheet" href="metropolis-fontsCustom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: middle, inverse, title-slide

.title[
# Single-cell RNA sequencing ~ Session 1
]
.subtitle[
## <html><br />
<br />
<hr color='#EB811B' size=1px width=796px><br />
</html><br />
Bioinformatics Resource Center - Rockefeller University
]
.author[
### <a href="https://rockefelleruniversity.github.io/SingleCell_Bootcamp/" class="uri">https://rockefelleruniversity.github.io/SingleCell_Bootcamp/</a>
]
.author[
### <a href="mailto:brc@rockefeller.edu" class="email">brc@rockefeller.edu</a>
]

---




class: inverse, center, middle

# Introduction to single cell sequencing

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---

## Overview

- High-throughput Sequencing
- Emergence of single-cell
- Differing single-cell assays
- Differing single-cell technologies
- Overview of Single-cell sequencing

---

## Sequencing

The emergence of high-throughput sequencing technologies such as 454 (Roche) and Solexa (Illumina) sequencing allowed for the highly parallel short read sequencing of DNA molecules.

- 454 (Roche) ~ 2005
- Solexa (Illumina)  ~ 2006

---
## Sequencing overview

![overview](imgs/wps.png)

---

## Sequence files - FastQ
 **FastQ (FASTA with Qualities)**
 
&lt;div align="center"&gt;
&lt;img src="imgs/fq1.png" alt="igv" height="200" width="800"&gt;
&lt;/div&gt;

- "@" followed by identifier.
- Sequence information.
- "+" 
- Quality scores encodes as ASCI.

---

## Sequence files - FastQ
 **FastQ - Header**
 
&lt;div align="center"&gt;
&lt;img src="imgs/fq2.png" alt="igv" height="200" width="800"&gt;
&lt;/div&gt;

- Header for each read can contain additional information
    - HS2000-887_89 - Machine name.
    - 5 - Flowcell lane.
    - /1 - Read 1 or 2 of pair (here read 1)
    
---

## Sequence files - FastQ

**FastQ - Qualities**

&lt;div align="center"&gt;
&lt;img src="imgs/fq3.png" alt="igv" height="200" width="800"&gt;
&lt;/div&gt;
 
- Qualities follow "+" line.
- -log10 probability of sequence base being wrong. 
- Encoded in ASCI to save space.
- Used in quality assessment and downstream analysis
 

---
## Sequencing use cases

- ChIP-seq - Study transcription factor binding
- RNA-seq - Study genome wide transcriptomics
- Whole Genome/Exome Sequencing - Study genome variation/changes
- MNAseq/DNAse-seq/ATAC-seq - Study chromatin structure/accessibility
- CrispR-seq - Study genome wide phenotypes
- And many more..

---
## Bulk sequencing methods

Sequencing typically performed on bulk tissue or cells.

- Tissues - Liver vs Heart
- Conditions - Cancer vs surrounding normal tissue
- Treatments - Hep-G2 cells with/without rapamycin 
- Genetic perturbation - Ikzf WT/KO in B-cells.

Analysis of the bulk characteristics of data without understanding of hetergeneity of data.

---
## FACS Cell and Nuclei sorting

Newer technologies such as TRAP from the Heintz lab or nuclei sorting allow for capture of distinct cell types based on expressed markers.

Pros
- Allow for the capture of rare cell populations such as specific neuron types.

Cons
- Require known markers for desired cell populations.

---
## Single-cell sequencing

With the advent of advanced microfluidics and refined sequencing technologies, single-cell sequencing has emerged as a technology to profile individual cells from a heterogeneous population without prior knowledge of cell populations.

Pros
- No prior knowledge of cell populations required.
- Simultaneously assess profiles of 1000s of cells.

Cons
- Low sequencing sequencing depth for individual cells (1000s vs millions of reads for bulk). 

---

## Single-cell use cases

Single-cell sequencing, as with bulk sequencing, has now been applied to the study of a wide range of differing assays.

- **scRNA-seq/snRNA-seq - Transcriptomics**
- scATAC-seq/snATAC-seq - Chromatin accessibility
- CITE-seq - Surface protein expression
- AIRR - TCR/BCR profiling
- Spatial-seq - Spatial profiling of gene expression.

---

## Single-cell platforms

- **10x Genomics (microfluidics)**
- Smart-seq3 (microfluidics)
- Drop-seq (microfluidics)
- VASA-seq (FANS)
- inDrop-seq (microfluidics)
- In-house solutions 


Many companies offer single-cell sequencing technologies which may be used with the Illumina sequencer.

---

## Major Single-cell platforms

Two popular major companies offer the most used technologies.

- **10x Genomics**
- Smart-seq3

Major difference between the two are the sequencing depth and coverage profiles across transcripts.

- 10x is 3' biased so will not provide full length transcript coverage.
- As 10x is sequencing only the 3' of transcripts/genes, less sequencing depth is required per transcript/gene


---
## The 10x Genomics system

![overview](imgs/CopyOf10x_full.png)

---
## The 10x Genomics system - Microfluidics

![Microfluidics](imgs/CopyOf10x_a.png)

---
## The 10x Genomics system  - Gel beads

![beads](./imgs/CopyOf10x_bead.png)

---

## The 10x Genomics system  - Sequencing primers

![primers](./imgs/CopyOf10x_sequencing.png)

---

## The 10x Genomics system  - Sequencing R1

**Read 1**

![read1](imgs/CopyOfRead1.png)

---
## The 10x Genomics system  - Sequencing R2

**Read 2**

![read2](imgs/CopyOfRead2.png)


---
## The 10x sequence reads

The sequence reads contain:- 

- Information on cell identity (10x barcode)
- Information on molecule identity (UMI)
- Information on sample identity (Index read)
- Information on the RNA molecule (transcriptome read)

---
## From Sequences to Processed data

As with standard bulk sequencing data, the next steps are typically to align the data to a reference genome/transcriptome and summarize data to a signal matrix.

- Align reads to transcriptome and summarize signal to genes for RNAseq
- Align reads to genome, call enriched windows/peaks and summarize signal in peaks.

---

## Processing scRNA-seq/snRNA-seq data.

For the processing of scRNA/snRNA from fastQ to count matrix, there are many options available to us.

**Alignment and counting**
- Cellranger count
- STAR - STARsolo
- Subread cellCounts

**Pseudoalignment and counting**
- Salmon - Alevin
- Kallisto - Bustools

---

## Droplets to counts

.pull-left[
The output of these tools is typically a matrix of the signal attributed to cells and genes (typically read counts).

This matrix is the input for all downstream post-processing, quality control, normalization, batch correction, clustering, dimension reduction and differential expression analysis.

The output matrix is often stored in a compressed format such as:- 
- MEX (Market Exchange Format)
- HDF5 (Hierarchical Data Format)


]

.pull-right[
![overview](imgs/countingDrops.png)
]

---

## HDF5 format

- Optimized format for big data.
- Contains matrix information alongside column and row information.
- Loom format version of HDF5 most popular.
- Efficient to work with programmatically.

![](https://linnarssonlab.org/loompy/_images/Loom_components.png)
---
## MEX format

- Plain/Compressed text format
- Contains matrix information in tsv file
- Separate row and column files contain barcode (cell) and feature (gene/transcript) information.

![](./imgs/matrixFile.png)

This is an example of a directory produced by Cell Ranger. 

---

class: inverse, center, middle

# Cell Ranger

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---


## Cell Ranger

Cell Ranger is the typical approach we use to process 10x data. The default setting are pretty good. This is an intensive program, so we will not be running this locally on your laptops. Instead we run it on remote systems , like the HPC.

If you are working with your own data, the data will often be provided as the Cell Ranger output by the Genomics/Bioinformatics teams, like here at Rockefeller University. 

---
## Cell Ranger

.pull-left[
Cell Ranger is a suite of tools for single cell processing and analysis available from 10X Genomics. It performs key processing steps i.e. demultiplexing, conversion to FASTQ and mapping. It is also the first chance to delve into your data sets QC. 

In this session we will give a brief overview of running this tool and then dive deeper into interpreting the outputs.
]

.pull-right[
![CellRanger](imgs/CopyOfCellRangerPage.png)
]

---
## Do I need to run this?

- Often genomics centers will run it for you and deliver mtx/hdf5 files (i.e. here at Rockefeller)

- Why run Cell Ranger yourself?:
  * You want to integrate data sets, but they were processed by different versions of Cell Ranger
  * You may want to change Cell Ranger parameters i.e. export BAMs, or force an expected cell number
  * Most commonly you want to reanalyze a published data set



---
## Cell Ranger Download

.pull-left[
- Cell Ranger is available from the [10x genomics website](https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest).

- Also available are pre-baked references for Human and Mouse genomes (GRCh37/38 and GRCm37)

]
.pull-right[![](./imgs/CellRangerDownload.png)
]

---
## Cell Ranger

- Cell Ranger only runs on linux machines (CentOS/RedHat 7.0+ and Ubuntu 14.04+).
- Due to memory requirements, typically users run Cell Ranger on a remote server and not their own machines
- To download Cell Ranger and the required reference onto a remote server, we typically use the *wget* command (these commands are all on the [Download](https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest) page)

[This will all be in terminal on the server you are using]


``` shell
wget -O cellranger-8.0.0.tar.gz "https://cf.10xgenomics.com/releases/cell-exp/cellranger-8.0.0.tar.gz?Expires=1711772964&amp;Key-Pair-Id=APKAI7S6A5RYOXBWRPDA&amp;Signature=muvzcbqxba6d-blyYS02MVfLlzwZk6iZNQWXdaoCLnl7owW2nEN-IHwSPwdNoYl-6Xia7rr0S1sLCUQTsekGm2pQKcd0kqK~ndHK0DM7SwSVpXLlRvBV5pXt~EIlsxATVBKVeQLnUy698N-WnRlT~ahjlU-nMdpomX9-lOkF~w8gbgHBdtPXunTWfW87sSJLpHMDVENSF7TFJsXERDwDnsXyQLCuEhfGTCOnupkaATlLEr9kaeCStePKkwGyqgi1m8Ua02NNGHWPIJ6I1mDt695wo~dgptpJF4SDNRTyE-TuXrHfIqRjZB60zhWRJczFo2kpL7FCKwliE-vJ6djcSw__"
```

Download reference genome for Cell Ranger i.e. Human genome (GRCh38)

``` shell
wget "https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-GRCh38-2024-A.tar.gz"
```


---
## Cell Ranger set-up

Having downloaded the software and references, we can then unpack them.


``` shell
tar -xzvf cellranger-8.0.0.tar.gz
tar -xzvf refdata-gex-GRCh38-2020-A.tar.gz
```

Finally we can add the cellranger directory to our PATH.


``` shell
export PATH=/PATH_TO_CELLRANGER_DIRECTORY/cellranger-8.0.0:$PATH
```

---
## Running Cell Ranger count

Now we have the downloaded Cell Ranger software and required pre-build reference for Human (GRCh38) we can start the generation of count data from scRNA-seq/snRNA-seq fastQ data.

Typically FastQ files for your scRNA run will have been generated using the **Cell Ranger mkfastq** toolset to produce a directory a FastQ files. 

As we have the downloaded Cell Ranger software and required pre-build reference for Human (GRCh38) we can generate count data from the FASTQ files. This will make a count matrix and associated files.

If you are analyzing single nuclei RNA-seq data remember to set the **--include-introns** flag.


``` shell
cellranger count --id=my_run_name \
   --fastqs=PATH_TO_FASTQ_DIRECTORY \
   --transcriptome=/PATH_TO_CELLRANGER_DIRECTORY/refdata-gex-GRCh38-2020-A
   --create-bam=true
```

---
## Working with custom genomes

If you are working with a genome which is not Human and/or mouse you will need to find another source for your Cell Ranger reference.

- Luckily many references are pre-built by other consortiums.
- We can build our own references using other tools in Cell Ranger using the *mkgtf* and *mkref* functions.
- For this you just need a FASTA file (DNA sequence) and a GTF file (Gene Annotation) for your reference.

---
## Creating your own reference

To create your own references you will need two additional files.

- FASTA file - File containing the full genome sequence for the reference of interest
- GTF file - File containing the gene/transcript models for the reference of interest.

---
## FASTA file (Genome sequence)

- The reference genome stored as a collection of contigs/chromosomes.
- A contig is a stretch of DNA sequence encoded as A,G,C,T,N.
- Typically comes in FASTA format.
    - "&gt;" line contains information on contig
    - Lines following contain contig sequence

&lt;div align="center"&gt;
&lt;img src="imgs/fasta.png" alt="igv" height="300" width="600"&gt;
&lt;/div&gt;

---
## GTF (Gene Transfer Format)

 &lt;div align="center"&gt;
&lt;img src="imgs/gff.png" alt="igv" height="150" width="700"&gt;
&lt;/div&gt;

- Used to genome annotation.
- Stores position, feature (exon) and meta-feature (transcript/gene) information.

- **Importantly for Cell Ranger Count, only features labelled as exon (_column 3_) will be considered for counting signal in genes**
- **Many genomes label mitochondrial genes with _CDS_ and not _exon_ so these must be updated**

---
## Using Cell Ranger mkgtf

Now we have the gene models in the GTF format we can use the **Cell Ranger mkgtf** tools to validate our GTF and remove any unwanted annotation types using the attribute flag.

Below is an example of how 10x generated the GTF for the Human reference.


``` shell
cellranger mkgtf Homo_sapiens.GRCh38.ensembl.gtf \
Homo_sapiens.GRCh38.ensembl.filtered.gtf \
                   --attribute=gene_biotype:protein_coding \
                   --attribute=gene_biotype:lncRNA \
                   --attribute=gene_biotype:antisense \
                   --attribute=gene_biotype:IG_LV_gene \
                   --attribute=gene_biotype:IG_V_gene \
                   --attribute=gene_biotype:IG_V_pseudogene \
                   --attribute=gene_biotype:IG_D_gene \
                   --attribute=gene_biotype:IG_J_gene \
                   --attribute=gene_biotype:IG_J_pseudogene \
                   --attribute=gene_biotype:IG_C_gene \
                   --attribute=gene_biotype:IG_C_pseudogene \
                   --attribute=gene_biotype:TR_V_gene \
                   --attribute=gene_biotype:TR_V_pseudogene \
                   --attribute=gene_biotype:TR_D_gene \
                   --attribute=gene_biotype:TR_J_gene \
                   --attribute=gene_biotype:TR_J_pseudogene \
                   --attribute=gene_biotype:TR_C_gene
```


---
## Using Cell Ranger mkref

Following filtering of your GTF to the required biotypes, we can use the Cell Ranger mkref tool to finally create our custom reference.


``` shell
cellranger mkref --genome=custom_reference \
--fasta=custom_reference.fa  \
--genes=custom_reference_filtered.gtf
```

---

class: inverse, center, middle

# Output files

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---



## Outputs

Having completed the Cell Ranger count step, the user will have created a folder, with the name set by the **--id** flag from the count command.

Within this folder there will be the **outs/** directory which contains all the outputs generated from Cell Ranger count.

![](./imgs/outputs.png)

---
## Count Matrices 

The count matrices to be used for further analysis are stored in both MEX and HDF5 formats within the output directories.

The filtered matrix only contains detected, cell-associated barcodes whereas the raw contains all barcodes (background and cell-associated).

**MEX format**
- filtered_feature_bc_matrix
- raw_feature_bc_matrix

**HDF5 format**
- filtered_feature_bc_matrix.h5
- raw_feature_bc_matrix.h5

---
## BAM files 

.pull-left[
The outs directory may also contain a BAM file of alignments for all barcodes against the reference _(possorted_genome_bam.bam)_ as well as an associated BAI index file _(possorted_genome_bam.bam.bai)_. This depends on whether you put true or false in the *--create-bam argument*. Older versions of Cell Ranger did not have this argument and would default to producing this BAM file. 

This BAM file is sometimes used in downstream analysis such as scSplit/Velocyto as well as for the generation of signal graphs such as bigWigs.
  ]

.pull-right[

&lt;div align="center"&gt;
&lt;img src="imgs/CopyOfsam1.png" alt="igv" height="200" width="450"&gt;
&lt;/div&gt;

&lt;div align="center"&gt;
&lt;img src="imgs/CopyOfsam2.png" alt="igv" height="150" width="450"&gt;
&lt;/div&gt;

]


---
## Cloupe files

Cell Ranger also outputs _cloupe.cloupe_ files for visualization within the [10X Loupe browser software](https://www.10xgenomics.com/support/software/loupe-browser/latest).

This allows for the visualization of scRNA-seq/snRNA-seq as a t-sne/umap with the ability to overlay metrics of QC and gene expression onto the cells in real time.

&lt;div align="center"&gt;
&lt;img src="imgs/CopyOfcloupe.png" alt="igv" height="350" width="400"&gt;
&lt;/div&gt;


---

class: inverse, center, middle

# Web Summary QC

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---

## QC outputs

Assessment of the overall quality of a scRNA-seq/snRNA-seq experiment after Cell Ranger can give our first chance to dig into the quality of your dataset and gain insight any issues we might face in data analysis.

---
## QC is essential

.pull-left[
There are many potential issues which can arise in scRNA-seq/snRNA-seq data including:

- Empty droplets
- Low quality cells (dead or dying)
- Ambient RNA contamination
- Doublets
  ]

.pull-right[

&lt;div align="center"&gt;
&lt;img src="imgs/CopyOfbad_cells.png" alt="igv" height="460" width="310"&gt;
&lt;caption&gt;Huemos et al. (2023)&lt;/caption&gt;
&lt;/div&gt;

  ]

---
## Metrics and Web Summary

Cell Ranger will also output summaries of useful metrics as a text file _(metrics_summary.csv)_ and as a intuitive web-page.

Metrics include:

- Counts/UMIs per cell.
- Number of cells detected.
- Alignment quality.
- Distribution of reads in genomic features.
- Sequencing saturation
- t-sne/UMAP with default clustering.

---
## Metrics and Web Summary

Assessment of the overall quality of a scRNA-seq/snRNA-seq experiment after Cell Ranger can give our first insight into any issues we might face.

---
## Web Summary Example

We will be looking at the web summary generated from a PBMC dataset with ~1,000 cells from a healthy human donor. The full experiment details can be found on the 10X website [here](https://www.10xgenomics.com/datasets/1-k-pbm-cs-from-a-healthy-donor-v-3-chemistry-3-standard-3-0-0) and you can get a copy of a web summary from Cell Ranger version 8 [here](./../../data/web_summary.html).

[NOTE: Older web summaries contain largely the same information, but slightly different layout]


---
## Web Summary overview

.pull-left[
The web summary html file contains an interactive report describing the most essential QC for your single cell experiment as well as initial clustering and dimension reduction for your data.

The web summary also contains useful information on the input files and the versions used in this analysis for later reproducibility.
]

.pull-right[
![websum](imgs/CopyOfwebsum1.png)
]

---
## Run Summary

.pull-left[
The first thing we can review is the **Run Summary** information panel. **(Top Left)**
As most people do not run Cell Ranger themselves this is important to check it matches expectations.

- Sample ID - Sample name (Assigned in **cellranger count**)
- Chemistry - The 10x chemistry used
- Include introns - Whether counting was run to include intronic counts (typical for single nucleus RNAseq, but also useful for scRNAseq)
- Pipeline Version - Version of Cell Ranger used

  ]

.pull-right[
![websum](imgs/CopyOfwebsum4.png)
  ]

---
## Command Line Arguments

A corresponding section of the web summary is the Command Line Arguments used to run Cell Ranger. **(Bottom)**
Again this is an important section to double-check to make sure everything was run correctly. 

- Input Path - Path to the input fastq files
- Reference Path and Transcriptome - References used in analysis

This is now under the Experimental Design tab of Cell Ranger 9. 


![websum](imgs/CopyOfwebsum5.png)


---
## Mapping panel

.pull-left[
The **Mapping** panel highlights information on the mapping of reads to the reference genome and transcriptome. **Bottom Left**

- Reads Mapped to Genome - Total mapped reads
- Reads Mapped Confidently to Genome - Uniquely mapped reads
- Reads Mapped Confidently to Exonic/Intronic/Intergenic - Uniquely mapped reads to specific regions
- Reads Mapped Confidently to Transcriptome - Reads mapped to a unique gene (and consistent with slice junctions)
- Reads Mapped Antisense to Gene - Reads mapped to the opposite strand of a gene
  ]
  
.pull-right[
![websum](imgs/CopyOfwebsum3.png)
]

---
## Mapping panel

Key Metrics we look for:

**Mapped to Genome &gt; 60% (usually range 50% ~ 90%)**

  * Mapping rate to reference genome
  * Check reference genome version if too low

**Reads Mapped Confidently to Transcriptome &gt; 30% (usually &gt; 60%)**

  * Reflection of annotation to transcriptome
  * Check annotation if too low



---
## Sequencing panel

.pull-left[
The **Sequencing** panel highlights information on the quality of the Illumina sequencing. **Top Right**.

- Number of reads - Total number of paired reads in library
- Number of Short Reads Skipped - Reads that were filtered out for being too short
- Valid Barcodes - Proportion of barcodes matching barcodes in whitelist (~ 740 thousand)
- Valid UMIs - Total number of UMIs that are not all one base and contain no unknown bases
- Sequencing saturation - Unique valid barcode/UMI versus all valid barcode/UMI
- Q30 scores - Assessment of sequencing qualities for barcode/UMI/RNA Reads

  ]

.pull-right[
![websum](imgs/CopyOfwebsum2.png)
]


---
## Sequencing panel

Key Metrics we look for:

**Q30 Bases in RNA Read &gt; 65% (usually &gt; 80%)**

  * Reflects the sequencing quality
  * Need to check with sequencing service supplier
		
**Sequencing Saturation &gt; 40% (usually range 20% ~ 80%)**

  * Reflects the complexity of libraries
  * Consider reconstructing library if too low


---
## Cells panel

.pull-left[
The **Cells** panel highlights some of the most important information in the report: the total number of cells captured and the distribution of counts across cells and genes. **Top Right**. Their importance is clear as several metrics are repeated and placed as the headline of the report. 

- Estimated Number of Cells - Total number of barcodes associated to at least one cell.
- Mean Reads per cell -  Average reads in each cell
- Median Genes per Cell - Median number of genes detected (at least 1 count) per cell associated barcodes

  ]
  
.pull-right[
![websum](imgs/CopyOfwebsum6.png)
]

---
## Cells panel 

.pull-left[
The **Cells** panel also has other metrics which help describe the depth and ambient RNA proportion. **Top Right**.

- Median Reads per Cell - Median number of transcriptome reads within cell associated barcodes
- Fraction Reads in Cells - Fraction of reads from valid barcode, associated to a cell and mapped to transcriptome
- Total Genes Detected - Number of genes with at least 1 count

  ]
  
.pull-right[
![websum](imgs/CopyOfwebsum6.png)
  ]


---
## Cells panel

Key Metrics we look for:

**Fraction Reads in Cells &gt; 70% (usually &gt; 85%)**

  * Reflects the ambient RNA contamination
  * Consider correcting for ambient RNA if &lt; 90%

**Median reads per cell &gt; 20,000/cell and estimated number of cells 500 - 10,000**

  * May be caused by the failure of cell identification
  * Need to check knee plot and re-evaluate cell number


---
## Knee plot

.pull-left[

The Cell panel also includes an interactive knee plot.

The knee plot shows:

- On the x-axis, the barcodes ordered by the most frequent on the left to the least frequent on the right

- On the y-axis, the frequency of each ordered barcode.

- Highlighted in dark blue are the barcodes marked as associated to cells.

]
.pull-right[
![](./imgs/CopyOfwebsum6.png)
]

---
## Knee plot

.pull-left[

It is apparent that barcodes labelled blue (cell-associated barcodes) do not have a cut-off just based on the UMI count.

In newer versions of Cell Ranger a two step process is used to define cell-associated barcodes based on the EmptyDrops method (Lun et al.,2019).

- First high RNA containing cells are identified based on a UMI cut-off.
- Second, low UMI containing cells are used as a background training set to identify additional cell-associated barcodes not called in first step.

If required, a **--force-cells** flag can be used with _cellranger count_ to identify a specific number of cell-associated barcodes.

  ]
  
.pull-right[
![websum](imgs/CopyOfwebsum6.png)
  ]

---
## Knee plot

The Knee plot also acts a good QC tools to investigate differing types of single cell failure.

Whereas our previous knee plot represented a good sample, differing knee plot patterns can be indicative of specific problems with the single cell protocol. We will show you some examples of these below from real data.


---
## Knee plot

.pull-left[
In this example we see no specific cliff and knee suggesting a failure in the integration of oil, beads and samples (wetting failure) or a compromised sample.
  ]
  
.pull-right[
![websum](imgs/CopyOfwetfailure.png)
  ]

---
## Knee plot

.pull-left[

If there is a clog in the machine we may see a knee plot where the overall number of samples is low.

  ]

.pull-right[
![websum](imgs/CopyOflowbarcode.png)
  ]

---
## Knee plot

.pull-left[

There may be occasions where we see two sets of cliff-and-knees in our knee plot.

This could be indicative of a heterogenous sample where we have two populations of cells with differing overall RNA levels.

Knee plots should be interpreted in the context of the biology under investigation.

  ]

.pull-right[
![websum](imgs/CopyOfhetero.png)
  ]
  
---
## Knee plot

It is important to know what version and parameters were used to run Cell Ranger. 

This cell calling step is continually updated and it can have a dramatic affect on your results. The [Web Summary](https://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_web_summary.html) on 10X Genomics for this dataset is from Cell Ranger V3.0 if you want to compare. 

Cell Ranger V9.0 was released relatively recently and, as with almost every prior version, there's been a change in default parameters (EmptyDrops false discovery rate (FDR) threshold has been lowered).


---
## Gene Expression page

.pull-left[
The web-summary also contains an analysis page where default dimension reduction, clustering and differential expressions between clusters has been performed.

Additionally the analysis page contains information on sequencing saturation and gene per cell vs reads per cell.

  ]

.pull-right[
![analysis](imgs/CopyOfanal1.png)
  ]

---
## t-SNE/UMAP and clustering

.pull-left[
The t-SNE plot shows the distribution and similarity within your data.

- Review for high and low UMI cells driving t-SNE structure and/or clustering.
- Expected separation between and structure across clusters may be observed within the t-SNE plot.
- Identify expected clusters based on expression of marker genes

  ]

.pull-right[
![analysis](imgs/CopyOfanal2.png)
  ]

---
## Sequence and Gene saturation

.pull-left[
The sequence saturation and Median genes per cell plots show these calculations (as show on summary page) over successive downsampling of the data. 

By reviewing the curve of the down sampled metrics we can assess whether we are approaching saturation for either of these metrics.

]

.pull-right[

&lt;div align="center"&gt;
&lt;img src="imgs/CopyOfanal3.png" alt="igv" height="200" width="200"&gt;
&lt;/div&gt;

&lt;div align="center"&gt;
&lt;img src="imgs/CopyOfanal4.png" alt="igv" height="200" width="200"&gt;
&lt;/div&gt;
]

---
## QC issues going forward?

Early issues with QC can manifest in many ways downstream. This is from a published dataset:

.pull-left[
![](./imgs/contaminated_knee.png)
  ]

.pull-right[
![](./imgs/marker_gene.png)
  ]
  
---
## Next steps

Fingers crossed there's no QC issues. 

Often at this step we wouldn't make any decisions unless there is a clear complete failure. This is an important first step in setting expectations/preparing for what you may need to do for the dataset.

---
## Cell Ranger 9

New Cell Ranger came out recently. We are yet to test it but new features include:

* Automated annotation for human data sets [in beta]
* UMAPs are made instead of tSNE
* New [Web Summary](https://cf.10xgenomics.com/samples/cell-exp/9.0.0/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_web_summary.html) layout
* EmptyDrops false discovery rate (FDR) threshold has been lowered 
* cellranger mkfastq is being deprecated
* [New Workflow](https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/running-pipelines/cr-3p-multi#hashing-5p-gex-vdj) for hashtagged data

  




---

class: inverse, center, middle

# Comparison of other scRNA-seq assays

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---


## Non-droplet based assays

.pull-left[

While the traditional 10X scRNA-seq transcriptomic assay is droplet based, other emerging methods are using non-droplet based methods to increase efficiency. 

For example, ScaleBio uses sample fixation, followed by library construction in which transcripts are sequentially barcoded using a specialized plate design. 

]

.pull-right[
![scalebio](imgs/CopyOfScaleBio_scRNA_Workflow-1-2048x635.png) 
]

---


## Non-droplet based assays

.pull-left[

- With non-droplet based methods, different QC metrics have to be considered. 

- For example, ambient RNA correction software is droplet-based, so you wouldn't use these methods for ScaleBio.


]




---
## How spatial works - Visium from 10X

.pull-left[

Fresh or frozen tissue is fixed onto a slide, and gene expression is captured from each section of the slide.

Each captured section has barcoded spots containing oligonucleotides with spatial barcodes unique to that spot.

Tissue is stained and then permeabilized, allowing released mRNA to bind to spatially barcoded oligonucleotides present on the spots. 

]
.pull-right[
![visium](imgs/CopyOfVisium-gene-expression-slide-1024x390.png) 
]

---
## How spatial works - Visium from 10X

.pull-left[

The outputs are filtered and raw matrices with gene expression information for each tissue-associated barcode.
 
Captured spots contain more than one cell, and need to be deconvoluted to predict celltype proportions in each spot.

You can use scRNA-seq data as a reference to estimate cellular proportion based on the gene expression profiles of each spot. 

There is available software to do this:
  - [BayesPrism](https://github.com/Danko-Lab/BayesPrism): Deconvolution 
  - [SpaceFold](https://github.com/dpeerlab/SpaceFold?tab=readme-ov-file): Mapping cells onto each spot

]

.pull-right[
![spatial](imgs/CopyOfSpatial-AnnotatedSCdata-Illus.png) 
]

    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
  "highlightStyle": "github",
  "highlightLines": true,
  "countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
