<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Single-cell RNA sequencing ~ Session 1 ¶  ¶ ¶  ¶ </title>
    <meta charset="utf-8" />
    <meta name="author" content="Rockefeller University, Bioinformatics Resource Centre" />
    <script src="libs/header-attrs-2.25/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="metropolisCustom.css" type="text/css" />
    <link rel="stylesheet" href="metropolis-fontsCustom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

.title[
# Single-cell RNA sequencing ~ Session 1 ¶ <html> ¶ ¶ <hr color='#EB811B' size=1px width=796px> ¶ </html>
]
.author[
### Rockefeller University, Bioinformatics Resource Centre
]
.date[
### <a href="http://rockefelleruniversity.github.io/SingleCell_Bootcamp/" class="uri">http://rockefelleruniversity.github.io/SingleCell_Bootcamp/</a>
]

---





class: inverse, center, middle

# Introduction to single cell sequencing

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---

## Overview

- High-throughput Sequencing
- Emergence of single-cell
- Differing single-cell assays
- Differing single-cell technologies
- Overview of Single-cell sequencing

---

## Sequencing

The emergence of high-throughput sequencing technologies such as 454 (Roche) and Solexa (Illumina) sequencing allowed for the highly parallel short read sequencing of DNA molecules.

- 454 (Roche) ~ 2005
- Solexa (Illumina)  ~ 2006

---
## Sequencing overview

![overview](./imgs/wps.png)

---

## Sequence files - FastQ
 **FastQ (FASTA with Qualities)**
&lt;div align="center"&gt;
&lt;img src="imgs/fq1.png" alt="igv" height="200" width="800"&gt;
&lt;/div&gt;

- "@" followed by identifier.
- Sequence information.
- "+" 
- Quality scores encodes as ASCI.

---

## Sequence files - FastQ
 **FastQ - Header**
 &lt;div align="center"&gt;
&lt;img src="imgs/fq2.png" alt="igv" height="200" width="800"&gt;
&lt;/div&gt;
- Header for each read can contain additional information
    - HS2000-887_89 - Machine name.
    - 5 - Flowcell lane.
    - /1 - Read 1 or 2 of pair (here read 1)
    
---

## Sequence files - FastQ
 **FastQ - Qualities**
 &lt;div align="center"&gt;
&lt;img src="imgs/fq3.png" alt="igv" height="200" width="800"&gt;
&lt;/div&gt;
 
- Qualities follow "+" line.
- -log10 probability of sequence base being wrong. 
- Encoded in ASCI to save space.
- Used in quality assessment and downstream analysis
 

---

## Sequencing use-cases

- ChIP-seq - Study transcription factor binding
- RNA-seq - Study genome wide transcriptomics
- Whole Genome/Exome Sequencing - Study genome variation/changes
- MNAseq/DNAse-seq/ATAC-seq - Study chromatin structure/accessibility
- CrispR-seq - Study genome wide phenotypes
- And many more..

---

## Bulk sequencing methods

Sequencing typically performed on bulk tissue or cells.

- Tissues - Liver vs Heart
- Conditions - Cancer vs surrounding normal tissue
- Treatments - Hep-G2 cells with/without rapamycin 
- Genetic perturbation - Ikzf WT/KO in B-cells.

Analysis of the bulk characteristics of data without understanding of hetergeneity of data.

---
## FACS Cell and Nuclei sorting

Newer technologies such as TRAP from the Heintz lab or nuclei sorting allow for capture of distinct cell types based on expressed markers.

Pros
- Allow for the capture of rare cell populations such as specific neuron types.

Cons
- Require known markers for desired cell populations

---
## Single-cell sequencing

With the advent of advanced microfluidics and refined sequencing technlogies, single-cell sequencing has emerged as a technology to profile individual cells from a heterogenous population without prior knowledge of cell populations.

Pros
- No prior knowledge of cell populations required.
- Simultaneously assess profiles of 1000s of cells.

Cons
- Low sequencing sequencing depth for individual cells (1000s vs millions of reads for bulk). 

---

## Single-cell use cases

Single-cell sequencing, as with bulk sequencing, has now been applied to the study of a wide range of differing assays.

- **scRNA-seq/snRNA-seq - Transcriptomics**
- scATAC-seq/snATAC-seq - Chromatin accessibility
- CITE-seq - Surface protein expression
- AIRR - TCR/BCR profiling
- Spatial-seq - Spatial profiling of gene expression.

---

## Single-cell platforms

- **10x Genomics (microfluidics)**
- Smart-seq3 (microfluidics)
- Drop-seq (microfluidics)
- VASA-seq	(FANS)
- inDrop-seq (microfluidics)
- In-house solutions 


Many companies offer single-cell sequencing technologies which may be used with the Illumina sequencer.

---

## Major Single-cell platforms

Two popular major companies offer the most used technologies.

- **10x Genomics**
- Smart-seq3

Major difference between the two are the sequencing depth and coverage profiles across transcripts.

- 10x is 3' biased so will not provide full length transcript coverage.
- As 10x is sequencing only the 3' of transcripts/genes, less sequencing depth is required per transcript/gene


---

## The 10x Genomics system

![overview](./imgs/10x_full.png)

---

## The 10x Genomics system - Microfluidics

![overview](./imgs/10x_a.png)

---

## The 10x Genomics system  - Gel beads

![overview](./imgs/10x_bead.png)

---

## The 10x Genomics system  - Sequencing primers

![overview](./imgs/10x_sequencing.png)

---

## The 10x Genomics system  - Sequencing R1

**Read 1**

![](./imgs/Read1.png)
---


## The 10x Genomics system  - Sequencing R2

**Read 2**

![](./imgs/Read2.png)
---

## The 10x Genomics system  - Sequence reads

The sequence reads contain:- 

- Information on cell identity (10x barcode)
- Information on molecule identity (UMI)
- Information on sample identity (Index read)
- Information on the RNA molecule (transcriptome read)

---

## From Sequences to Processed data

As with standard bulk sequencing data, the next steps are typically to align the data to a reference genome/transcriptome and summarise data to a signal matrix.

- Align reads to transcriptome and summarise signal to genes for RNAseq
- Align reads to genome, call enriched windows/peaks and summarise signal in peaks.

---

## Processing scRNA-seq/snRNA-seq data.

For the processing of scRNA/snRNA from fastQ to count matrix, there are many options available to us.

**Alignment and counting**
- Cellranger count
- STAR - STARsolo
- Subread cellCounts

**Pseudoalignment and counting**
- Salmon - Alevin
- Kallisto - Bustools

---

## Droplets to counts

.pull-left[
The output of these tools is typically a matrix of the signal attributed to cells and genes (typically read counts).

This matrix is the input for all downstream post-processing, quality control, normalisation, batch correction, clustering, dimension reduction and differential expression analysis.

The output matrix is often stored in a compressed format such as:- 
- MEX (Market Exchange Format)
- HDF5 (Hierarchical Data Format)


]
.pull-right[
![overview](./imgs/countingDrops.png)
]

---

## HDF5 format

- Optimized format for big data.
- Contains matrix information alongside column and row information.
- Loom format version of HDF5 most popular.
- Efficient to work with programmatically.

![](https://linnarssonlab.org/loompy/_images/Loom_components.png)

---

## MEX format

- Plain/Compressed text format
- Contains matrix information in tsv file
- Separate row and column files contain barcode (cell) and feature (gene/transcript) information.

![](./imgs/matrixFile.png)
---

class: inverse, center, middle

# Cell Ranger - Setting up

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---

## Cell Ranger

.pull-left[
Cell Ranger is a suite of tools for single cell processing and analysis available from 10x Genomics.

In this session we will make use of Cell Ranger **Count** tool to process our generated fastQ and create the required files.
]
.pull-right[
![](./imgs/CellRangerPage.png)
]

---
## Cell Ranger Download

.pull-left[
- Cell Ranger is available from the [10x genomics website](https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest).

- Also available are pre-baked references for Human and Mouse genomes (GRCh37/38 and GRCm37)

]
.pull-right[![](./imgs/CellRangerDownload.png)
]

---
## Cell Ranger Download page

&lt;iframe width="100%" height= "80%" src="https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest?"&gt;&lt;/iframe&gt;

---

## Cell Ranger

- Cell Ranger only runs on linux machines (CentOS/RedHat 7.0+ and Ubuntu 14.04+).
- Typically, due to memory requirements, users run Cell Ranger on a remote server and not their own machines.
- To download Cell Ranger and the required reference on remote server, we typically use the wget command.

Download the software
```{}
wget -O cellranger-7.1.0.tar.gz "https://cf.10xgenomics.com/releases/cell-exp/cellranger-7.1.0.tar.gz?Expires=1686030213&amp;Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiaHR0cHM6Ly9jZi4xMHhnZW5vbWljcy5jb20vcmVsZWFzZXMvY2VsbC1leHAvY2VsbHJhbmdlci03LjEuMC50YXIuZ3oiLCJDb25kaXRpb24iOnsiRGF0ZUxlc3NUaGFuIjp7IkFXUzpFcG9jaFRpbWUiOjE2ODYwMzAyMTN9fX1dfQ__&amp;Signature=jlyPxMYCCjFJZzrVMPcb7GSVZSYCCbyfnTQO2PlnCszG-ycgpHgUllHuV6l0ke7p5fPgM~m8xiQPiq-5VqDwdXKuGuKXtWMFFtakYnSroj7O79gQf2lKEXpRQlfeou5EEP4KQBjquwfbHZWs-NNFKyGMYjYrt6qxoyMzrcrgl2rEvRO7Pu8vwk0DJFnwRRRu~wFJEaDqUJ4vFXuKw1jT9aus~bzLeF4fsWDVQYfA7H71yc5zBvKxz1tfD-zTm7ARaA6j-gyC3ffQf9K5W7HSMJD8Iqez39-B8SHMwzsBv0o~uPlatSv-1YataeSHQQykRWxjZdrMg-5IL2neGrM8zA__&amp;Key-Pair-Id=APKAI7S6A5RYOXBWRPDA"
```

Download reference for Human genome (GRCh38)
```{}
wget -O https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-GRCh38-2020-A.tar.gz
```

---

## Cell Ranger set-up

- Having downloaded the software and references, we can then unpack them.

Unpack software and references
```{}
tar -xzvf cellranger-7.1.0.tar.gz
tar -xzvf refdata-gex-GRCh38-2020-A.tar.gz
```

- Finally we can add the cellranger directory to our PATH.

```{}
 export PATH=/PATH_TO_CELLRANGER_DIRECTORY/cellranger-7.1.0:$PATH
```

---

## Running Cell Ranger count

Now we have the downloaded Cell Ranger software and required pre-build reference for Human (GRCh38) we can start the generation of count data from scRNA-seq/snRNA-seq fastQ data.

Typically FastQ files for your scRNA run will have been generated using the **Cell Ranger mkfastq** toolset to produce a directory a fastQ files. 

We can now use CellRanger count command with our reference and fastQ files to generate our count matrix and associated files.

If you are analysing single nuclei RNA-seq data remember to set the **--include-introns** flag.

```{}
cellranger count --id=my_run_name \
   --fastqs=PATH_TO_FASTQ_DIRECTORY \
   --transcriptome=/PATH_TO_CELLRANGER_DIRECTORY/refdata-gex-GRCh38-2020-A
```
---

## Working with custom genomes

If you are working with a genome which is not Human and/or mouse you will need to find another source for your Cell Ranger reference.

- Luckily many references are pre-built by other consortiums.
- We can build our own references using other tools in Cell Ranger.

---

## Creating your own Cell Ranger references

To create your own references you will need two additional files.

- FASTA file - File containing the full genome sequence for the reference of interest
- GTF file - File containing the gene/transcript models for the reference of interest.

---

## FASTA file (Genome sequence)

- The reference genome stored as a collection of contigs/chromosomes.
- A contig is a stretch of DNA sequence encoded as A,G,C,T,N.
- Typically comes in FASTA format.
    - "&gt;" line contains information on contig
    - Lines following contain contig sequence

&lt;div align="center"&gt;
&lt;img src="imgs/fasta.png" alt="igv" height="300" width="600"&gt;
&lt;/div&gt;

---

## GTF (Gene Transfer Format)

 &lt;div align="center"&gt;
&lt;img src="imgs/gff.png" alt="igv" height="150" width="700"&gt;
&lt;/div&gt;

- Used to genome annotation.
- Stores position, feature (exon) and meta-feature (transcript/gene) information.

- **Importantly for Cell Ranger Count, only features labelled as exon (_column 3_) will be considered for counting signal in genes**
- **Many genomes label mitochondrial genes with _CDS_ and not _exon_ so these must be updated**

---

## Using Cell Ranger mkgtf

Now we have the gene models in the GTF format we can use the **Cell Ranger mkgtf** tools to validate our GTF and remove any unwanted annotation types using the attribute flag.

Below is an example of how 10x generated the GTF for the Human reference.

```{}
cellranger mkgtf Homo_sapiens.GRCh38.ensembl.gtf \
Homo_sapiens.GRCh38.ensembl.filtered.gtf \
                   --attribute=gene_biotype:protein_coding \
                   --attribute=gene_biotype:lncRNA \
                   --attribute=gene_biotype:antisense \
                   --attribute=gene_biotype:IG_LV_gene \
                   --attribute=gene_biotype:IG_V_gene \
                   --attribute=gene_biotype:IG_V_pseudogene \
                   --attribute=gene_biotype:IG_D_gene \
                   --attribute=gene_biotype:IG_J_gene \
                   --attribute=gene_biotype:IG_J_pseudogene \
                   --attribute=gene_biotype:IG_C_gene \
                   --attribute=gene_biotype:IG_C_pseudogene \
                   --attribute=gene_biotype:TR_V_gene \
                   --attribute=gene_biotype:TR_V_pseudogene \
                   --attribute=gene_biotype:TR_D_gene \
                   --attribute=gene_biotype:TR_J_gene \
                   --attribute=gene_biotype:TR_J_pseudogene \
                   --attribute=gene_biotype:TR_C_gene
```




---

## Using Cell Ranger mkref

Following filtering of your GTF to the required biotypes, we can use the Cell Ranger mkref tool to finally create our custom reference.

```{}
cellranger mkref --genome=custom_reference \
--fasta=custom_reference.fa  \
--genes=custom_reference_filtered.gtf
```

---

class: inverse, center, middle

# Cell Ranger -  Output files

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---


## Cell Ranger - Outputs

Having completed the Cell Ranger count step, the user will have created a folder named as set by the **--id** flag for count command.

Within this folder will be the **outs/** directory containing all the outputs generated from Cell Ranger count.

![](./imgs/outputs.png)

---

## Cell Ranger - Count Matrices 

The count matrices to be used for further analysis are stored in both MEX and HDF5 formats within the output directories.

The filtered matrix only contains detected, cell-associated barcodes whereas the raw contains all barcodes (background and cell-associated).

**MEX format**
- filtered_feature_bc_matrix
- raw_feature_bc_matrix

**HDF5 format**
- filtered_feature_bc_matrix.h5
- raw_feature_bc_matrix.h5

---

## Cell Ranger - BAM files 

.pull-left[
The outs directory also contains a BAM file of alignments for all barcodes against the reference _(possorted_genome_bam.bam)_ as well as an associated BAI index file _(possorted_genome_bam.bam.bai)_.

This BAM file is often used in downstream analysis such as scSplit/Velocyto as well as for the generation of signal graphs such as bigWigs.
]
.pull-right[
 &lt;div align="center"&gt;
&lt;img src="imgs/sam1.png" alt="igv" height="200" width="100"&gt;
&lt;/div&gt;
 &lt;div align="center"&gt;
&lt;img src="imgs/sam2.png" alt="igv" height="150" width="450"&gt;
&lt;/div&gt;
&lt;!-- ![](./) --&gt;
&lt;!-- ![](./imgs/sam2.png) --&gt;
]


---
## Cell Ranger - Cloupe files 

Cell Ranger also outputs files for visualisation within its own cloupe browser - _cloupe.cloupe_.

This allows for the visualisation of scRNA-seq/snRNA-seq as a t-sne/umap with the ability to overlay metrics of QC and gene expression onto the cells in real time

 &lt;div align="center"&gt;
&lt;img src="imgs/cloupe.png" alt="igv" height="400" width="450"&gt;
&lt;/div&gt;


---
## Cell Ranger - Metrics and Web-summary files

Cell Ranger will also output summaries of useful metrics as a text file _(metrics_summary.csv)_ and as a intuitive web-page.

Metrics include

- Counts/UMIs per cell.
- Number of cells detected.
- Alignment quality.
- Distibution of reads in genomic features.
- Sequencing saturation
- t-sne/UMAP with default clustering.

---
## QC is essential.

.pull-left[
There are many potential issues which can arise in scRNA-seq/snRNA-seq data including -

- Empty droplets.
- Low quality cell (dead or dying)
- Ambient RNA contamination.
- Doublet detection
]
.pull-right[
 &lt;div align="center"&gt;
&lt;img src="imgs/bad_cells.png" alt="igv" height="300" width="200"&gt;
&lt;/div&gt;
]

Assessment of the overall quality of a scRNA-seq/snRNA-seq experiment and filtering of low quality or contaminated cell counts is an essential step in analysis.

---

class: inverse, center, middle

# Cell Ranger - Web Summary QC

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---


## Web Summary overview

.pull-left[
The web summary html file contains an interactive report describing the most essential QC for your single cell experiment as well as initial clustering and dimension reduction for your data.

The web summary also contains useful information on the input files and the versions used in this analysis for later reproducibility.
]
.pull-right[
![](./imgs/websum1.png)
]

---
## Sample panel

.pull-left[
The first thing we can review is the **Sample** information panel.

This contains information on:-

- Sample ID - Sample name (Assigned in **cellranger count**)
- Chemistry - The 10x chemistry used
- Include introns - Whether counting was run to include intron counts (typical for single neuron RNA-seq).
- Reference Path and Transcriptome - References used in analysis
- Pipeline Version - Version of Cell Ranger used.


]
.pull-right[
![](./imgs/websum4.png)
]
---

## Sequencing panel

.pull-left[
The **Sequencing** panel highlights information on the quality of the illumina sequencing.

This contains information on:-

- Number of reads - Total number of paired reads in library
- Valid barcodes - Number of barcodes matching barcodes in whitelist (known to be kit ~ 1 million).
- Valid UMIs - Total number of UMIs that are not all one base and contain no unknown bases.
- Sequencing saturation - Unique valid barcode/UMI versus all valid barcode/UMI
- Q30 scores - Assessment of sequencing qualities for barcode/umi/index/RNA reads.


]
.pull-right[
![](./imgs/websum2.png)
]
---

## Mapping panel

.pull-left[
The **Mapping** panel highlights information on the mapping of reads to the reference genome and transcriptome.

This contains information on:-

- Reads mapped to genome - Total number of reads mapping to the genome
- Reads mapped confidently to genome - Reads mapping uniquely to the genome
- Reads mapped confidently to exonic/Intronic - Reads mapping uniquely to the exons or introns
- Reads mapped confidently to transcriptome - Reads mapping to a unique gene (and consistent with slice junctions).


]
.pull-right[
![](./imgs/websum3.png)
]
---

## Cells panel

.pull-left[
The **Cells** panel highlights some of the most important information in the report, the total number of cells captured and the distribution of counts across cells and genes.

Information includes:-

- Estimated number of cells - Total number of barcodes associated to at least one cell.
- Fraction reads in cells - Fraction of reads from valid barcode, associated to a cell and mapped to transcriptome. 
- Median reads per cell - Median number of transcriptome reads within cell associated barcodes
- Median genes per cell - - Median number of genes detected (at least 1 count) per cell associated barcodes.


]
.pull-right[
![](./imgs/websum6.png)
]
---


## Knee plot

.pull-left[

The Cell panel also includes and interactive knee plot.

The knee plot shows:- 

- On the x-axis, the barcodes ordered by the most frequent on the left to the least frequent on the right

- On the y-axis, the frequency of each ordered barcode.

- Highlighted in blue are the barcodes marked as associated to cells.



]
.pull-right[
![](./imgs/websum6.png)
]
---


## Knee plot

.pull-left[

It is apparent that barcodes labelled blue (cell-associated barcodes) do not have a cut-off based on UMI count.

In the latest version of Cell Ranger a two step process is used to define cell-associated barcodes based on the EmptyDrops method (Lun et al.,2019).

- First high RNA containing cells are identified based on a UMI cut-off.
- Second, low UMI containing cells are used as a background training set to identify additonal cell-associated barcodes not called in first step.

If required, a **--force-cells** flag can be used with _cellranger count_ to identify a set number of cell-associated barcodes.




]
.pull-right[
![](./imgs/websum6.png)
]
---

## Knee plot

.pull-left[

The Knee plot also acts a good QC tools to investigate differing types of single cell failure.

Whereas our previous knee plot represented a good sample, differing knee plot patterns can be indicative of specific problems with the single cell protocol.

In this example we see no specific cliff and knee suggesting a failure in the integration of oil, beads and samples (wetting failure) or a compromised sample.



]
.pull-right[
![](./imgs/wetfailure.png)
]
---

## Knee plot

.pull-left[

If there is a clog in the machine we may see a knee plot where the overall number of samples is low.



]
.pull-right[
![](./imgs/lowbarcode.png)
]
---

## Knee plot

.pull-left[

There may be occasions where we see two sets of cliff-and-knees in our knee plot.

This could be indicative of a heterogenous sample where we have two populations of cells with differing overall RNA levels.

Knee plots should be interpreted in the context of the biology under investigation.


]
.pull-right[
![](./imgs/hetero.png)
]
---


## Analysis page

.pull-left[
The web-summary also contains an analysis page where default dimension reduction, clustering and differential expressions between clusters has been performed.

Additionally the analysis page contains information on sequencing saturation and gene per cell vs reads per cell.

]

.pull-right[
![](./imgs/anal1.png)
]

---


## t-sne and clustering

.pull-left[
The t-sne plot shows the distribution and similarity within your data.

- Review for high and low UMI cells driving t-sne structure and/or clustering.
- Expected separation between and structure across clusters may be observed within the t-sne plot.
- Identify expected clusters based on expression of marker genes

]

.pull-right[
![](./imgs/anal2.png)
]

---


## Sequence and Gene saturation

.pull-left[
The sequence saturation and Median genes per cell plots show these calculations (as show on summary page) over successive downsampling of the data. 

By reviewing the curve of the downsamlped metrics we can assess whether we are approaching saturation for either of these metrics.

]

.pull-right[
 &lt;div align="center"&gt;
&lt;img src="imgs/anal3.png" alt="igv" height="200" width="200"&gt;
&lt;/div&gt;
 &lt;div align="center"&gt;
&lt;img src="imgs/anal4.png" alt="igv" height="200" width="200"&gt;
&lt;/div&gt;
]
---


class: inverse, center, middle

# Loupe Browser

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---

## Loupe Browser for visualization


.pull-left[

The Loupe browser is a tool for visualization of Cell Ranger cloupe files. 

It provides t-sne visualization of your single-cell data alongside sample/cell information as well as methods to test and visualise changes in gene expression.

Loupe browser can be freely downloaded from the 10x website.

]

.pull-right[
 &lt;div align="center"&gt;
&lt;img src="imgs/loupeDownload.png" alt="igv" height="400" width="350"&gt;
&lt;/div&gt;
]
---

## Loupe Browser for visualization


.pull-left[

Having downloaded the Loupe browser we can load our cloupe files directly in and rapidly visually interrogate our data.

In todays session we will review some of the features available in Loupe using the PBMC example data set.

You can download this dataset from our [S3 bucket](https://rubioinformatics.s3.amazonaws.com/cloupe.cloupe) to use in the Loupe browser

]

.pull-right[
 &lt;div align="center"&gt;
&lt;img src="imgs/loupeFront.png" alt="igv" height="300" width="400"&gt;
&lt;/div&gt;
]
---
## Resources

- [10X Single cell software](https://support.10xgenomics.com/single-cell-gene-expression/software/overview/welcome)
- [Cell Ranger Download](https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest)
- [Loupe Broswer Download](https://www.10xgenomics.com/products/loupe-browser/downloads)
- [Interpreting Web Summaries](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/summary)













    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
